lib/puppet/indirector/catalog/static_compiler.rb

```diff
@@ -91,8 +91,11 @@ def find_and_replace_metadata(request, resource, file)
   # @param resource [Puppet::Resource] The resource to add the metadata to
   # @param metadata [Puppet::FileServing::Metadata] The metadata of the given fileserver based file
   def replace_metadata(request, resource, metadata)
-    [:mode, :owner, :group].each do |param|
-      resource[param] ||= metadata.send(param)
+    file = resource.to_ral
+    if file.parameter(:source_permissions).value == :use
```

Comment:
Sadly the source_permissions logic is a bit of mess. The client sends the `source_permissions` as an HTTP query parameter to the fileserver, and the fileserver either stat's the files or provides default values based on the current process (unix) or sane defaults (windows). The fileserver copies the values into the catalog, which means the agent can't differentiate between "it was set explicitly in the manifest" vs "the fileserver injected values from files on disk" vs "the fileserver injected default values". Then when the agent applies the catalog, it sometimes ignores the metadata, e.g. we ignore owner/group when not root, we skip certain behavior on Windows, etc.

I think it makes sense in an ideal world to only inline metadata when `source_permissions` is `use` or `use_when_creating` (and when the parameter is not explicitly managed in the manifest). But it would mean a statically compiled catalog would have a different effect when applied than a "normal" catalog. That is, when applying a static catalog where `source_permissions => use`, we would not inline metadata for unmanaged parameters, so when a root agent applies the catalog, it would rely on the process umask. But when the agent applies a regular catalog, it would query the server for metadata, and copy the source permissions to the local file, even for permissions that were originally unmanaged.

I think the behavior you are proposing is more correct (with the caveat we should handle `use_when_creating` also), but I'm concerned about having different behavior in static vs non-static catalogs.